Add support for Direct Secret reference

baijum opened the pull request:
Fixes #872

baijum commented on pkg/reconcile/pipeline/context/service.go:
```diff
@@ -49,6 +50,17 @@ type service struct {
 }
 
+func (s *service) IsDirectSecretReference() bool {
+	if s.serviceRef.Kind == "Secret" && s.serviceRef.Version == "v1" {
```
group should be empty

baijum replied:
Done. 1bf562899688f37d3b28faaa2b8ca8c680441719

baijum replied:
I had to change commit because if DCO.  This is the new commit: 273c29d7b6ae20431f7c54d534e4ecddcdb16b8e

baijum commented on pkg/reconcile/pipeline/handler/collect/impl.go:
```diff
@@ -131,6 +132,22 @@ func ProvisionedService(ctx pipeline.Context) {
+
+	for _, service := range services {
+		if service.IsDirectSecretReference() {
+			res := service.Resource()
```
Extract secret directly from the object

baijum replied:
done. d11782fdf7ece95cffe1f7ac009f10850ef4964b

baijum commented on pkg/reconcile/pipeline/handler/collect/impl.go:
```diff
@@ -131,6 +132,22 @@ func ProvisionedService(ctx pipeline.Context) {
 	}
 }
 
+func DirectSecretReference(ctx pipeline.Context) {
```
Unit test for this function

baijum replied:
Done. fee11f2015b720a1d5984d357d438f8853a40e43

baijum commented on test/acceptance/features/bindAppToServiceUsingSecret.feature:
```diff
@@ -761,3 +761,44 @@ Feature: Bind values from a secret referred in backing service resource
             bla
             """
+
+    Scenario: Bind application through direct secret reference
```
Inject binding to an application from a Secret resource referred as service

baijum replied:
done. d11782fdf7ece95cffe1f7ac009f10850ef4964b

isutton commented on pkg/reconcile/pipeline/handler/collect/impl.go:
```diff
@@ -131,6 +132,23 @@ func ProvisionedService(ctx pipeline.Context) {
 }
 
+func DirectSecretReference(ctx pipeline.Context) {
+	services, _ := ctx.Services()
```
I see `ctx.Services()` implementation might return errors, any special reason this is being ignored?

baijum replied:
Fixed. 58aea3667d70a507d6f3f32d6ea0dda8d26cb465

isutton commented on pkg/reconcile/pipeline/handler/collect/impl.go:
```diff
@@ -131,6 +132,23 @@ func ProvisionedService(ctx pipeline.Context) {
+			name := res.GetName()
+			secret, err := ctx.ReadSecret(res.GetNamespace(), name)
+			if err != nil {
+				requestRetry(ctx, ErrorReadingSecret, err)
```
The request will be retried infinitely in the case the secret does not exist; `err` must be checked and retries shouldn't be requested in unrecoverable cases.

baijum replied:
What if someone creates a secret after the ServiceBinding resource creation?
I can see we have similar logic in ProvisionedService also. line 111-112.

isutton replied:
Kubernetes offers mechanisms to react on creation, update and deletion of resources to avoid situations such as the indicated. The appropriated approach would be a) fail without retry if a Secret is not found at the time of the reconciliation and b) observe the life-cycle (either fully or partially) of Secrets to perform the reconciliation of the related ServiceBinding.

baijum replied:
As I understand we removed all watches as part of optimization. So, I am not sure how to reconcile it.
As I pointed out, this design approach is there in other parts of our codebase.

isutton commented on pkg/reconcile/pipeline/handler/collect/impl.go:
```diff
@@ -131,6 +132,27 @@ func ProvisionedService(ctx pipeline.Context) {
+func DirectSecretReference(ctx pipeline.Context) {
+	services, err := ctx.Services()
+	if err != nil {
+		requestRetry(ctx, ErrorReadingSecret, err)
```
The error might not be related to reading a secret in this particular case.

baijum replied:
Fixed. cf2da0c816f48b61546bb862397dace9fb581ec6

isutton commented on test/acceptance/features/bindAppToServiceUsingSecret.feature:
```diff
@@ -761,3 +761,44 @@ Feature: Bind values from a secret referred in backing service resource
             bla
             """
+
+    Scenario: Inject binding to an application from a Secret resource referred as service
```
Given that it is expected the operator to recover from situations where the referred Secret comes to existence after the ServiceBinding resource is created, it'd be nice to see a reproduction of this scenario in the acceptance test suite.

baijum replied:
Done. 494077eab6dd88dd5b5db0a7060bf41b78d9fddf

isutton reviewed (approved):
LGTM

isutton commented on test/acceptance/features/bindAppToServiceUsingSecret.feature:
```diff
@@ -761,3 +761,86 @@ Feature: Bind values from a secret referred in backing service resource
+            """
+
+        Then Service Binding "bind-provisioned-service-1" is ready
+        And jq ".status.secret" of Service Binding "bind-provisioned-service-1" should be changed to "provisioned-secret-1"
```
Is this the expected behavior?

Now I wonder how this works with multiple secrets, given the ServiceBinding contains only one `.status.secret` field.

I believe we need an acceptance test for that scenario as well.

baijum replied:
Updated the current test. a5aedb0121563ceed50f7e2210be631620a08ba3

pedjak commented on pkg/reconcile/pipeline/handler/collect/impl.go:
```diff
@@ -131,6 +132,27 @@ func ProvisionedService(ctx pipeline.Context) {
 }
 
+func DirectSecretReference(ctx pipeline.Context) {
+	services, err := ctx.Services()
+	if err != nil {
+		requestRetry(ctx, ErrorReadingServicesReason, err)
+		return
+	}
```
You can omit this here, because we have this check already in PreFlight stage:

https://github.com/redhat-developer/service-binding-operator/blob/master/pkg/reconcile/pipeline/handler/collect/impl.go#L29

This stage was created to perform common checks for all followup stages.

baijum replied:
Done. 10cf41199e4e0a45796f32d33108c79969cac1f7

pedjak commented on pkg/reconcile/pipeline/handler/collect/impl_test.go:
```diff
@@ -745,6 +746,111 @@ var _ = Describe("Collect From Provisioned Service", func() {
+		})
+	})
+
+	Describe("unsuccessful processing", func() {
+		var (
+			services []pipeline.Service
+
+			err = errors.New("e")
+		)
+
+		BeforeEach(func() {
+			services = []pipeline.Service{}
+			ctx.EXPECT().Services().DoAndReturn(func() ([]pipeline.Service, error) { return services, err })
+		})
+		It("should retry processing if service listing return error", func() {
+			ctx.EXPECT().RetryProcessing(err)
+			ctx.EXPECT().SetCondition(v1alpha1.Conditions().NotCollectionReady().Reason(collect.ErrorReadingServicesReason).Msg(err.Error()).Build())
+
+			collect.DirectSecretReference(ctx)
+		})
+
+	})
+})
```
see above, this tests can be removed.

baijum replied:
Done. 10cf41199e4e0a45796f32d33108c79969cac1f7

pedjak commented on test/acceptance/features/bindAppToServiceUsingSecret.feature:
```diff
@@ -761,3 +761,110 @@ Feature: Bind values from a secret referred in backing service resource
+                version: v1
+                resource: deployments
+          """
+        * The Secret is present
```
In order to really validate this scenario, service binding must be detected as not ready before injecting secrets.

baijum replied:
Done. 10cf41199e4e0a45796f32d33108c79969cac1f7

pedjak replied:
I am not seeing any new line in this scenario checking that service binding is not ready.

baijum replied:
Please check line no. 876
Let me know if that is sufficient.

pedjak commented on test/acceptance/features/bindAppToServiceUsingSecret.feature:
```diff
@@ -761,3 +761,110 @@ Feature: Bind values from a secret referred in backing service resource
+                kind: Secret
+                name: provisioned-secret-1
+                id: one
+              - group: ""
```
do we really need two services for this scenario? If not let's keep the setup minimal.

baijum replied:
There was a comment from Igor about what happens if there are multiple Secret resources.
I tried to combine it. Looks like it's confusing, I will create a separate test case for that.

pedjak commented on test/acceptance/features/steps/steps.py:
```diff
@@ -249,7 +249,7 @@ def sbo_jq_is(context, jq_expression, sbr_name, json_value):
     polling2.poll(lambda: json.loads(
         openshift.get_resource_info_by_jq("servicebinding", sbr_name, context.namespace.name, jq_expression,
-                                          wait=False)) == json_value, step=5, timeout=800,
+                                          wait=False)).startswith(json_value), step=5, timeout=800,
```
this is going to result in a way more positive results - why is this change needed in this PR?

baijum replied:
I changed it to check the value of `.status.secret` when there are multiple Secret resources.
I will remove this change and create a separate step function for that requirement.

pedjak replied:
Where there are multiple secret referred, then the binding secret will be generated as before, so no need to check anything new there.

baijum replied:
Yes, but there was no assertion for the `.status.secret` value which is dynamic in nature. Please take a look at the test that I added and let me know.

pedjak replied:
in case when intermediate secret is generated and then injected, we should asset the following things:

* service binding Ready condition is true
* `.status.binding.name` is set, i.e. non empty
* the application sees injected env vars or files

The fact that intermediate secret name is made as `binding name` + `some hash` is an implementation detail and the user should not count on that - we can change that at any point in time.

baijum replied:
Done. 832c72fcec87d9f227ff5ea071ff759a112edc95

pedjak reviewed (changes requested):
we need  additional acceptance test demonstrating  additional bindings can be added on top of direct reference using `mappings`.

also, we support already using secret as reference, via annotations. Check this scenario:

https://github.com/redhat-developer/service-binding-operator/blob/master/test/acceptance/features/bindAppToService.feature#L796

How we are going to differentiate? Should we treat a secret as direct reference only if it does not contain annotations?

baijum commented:
> How we are going to differentiate? Should we treat a secret as direct reference only if it does not contain annotations?

I am fine with this approach.  I can add check for the annotation.

baijum commented on pkg/reconcile/pipeline/handler/collect/impl_test.go:
```diff
@@ -745,6 +746,111 @@ var _ = Describe("Collect From Provisioned Service", func() {
+			content.SetGroupVersionKind(schema.GroupVersionKind{Group: "", Version: "v1", Kind: "Secret"})
+			content.SetName(secretName1)
+
+			ctx.EXPECT().ReadSecret(ns1, secretName1).Return(content, err)
```
Done. 10cf41199e4e0a45796f32d33108c79969cac1f7

pedjak commented on test/acceptance/features/bindAppToServiceUsingSecret.feature:
```diff
@@ -761,3 +761,202 @@ Feature: Bind values from a secret referred in backing service resource
+                  value: '{{ .username }}:{{ .password }}'
+          """
+        Then Service Binding "bind-provisioned-service-1" is ready
+        And jq ".status.secret" of Service Binding "bind-provisioned-service-1" should starts with "bind-provisioned-service-1"
```
we should not advertise this kind of info to users - it is not a part of api/contract - the spec only says that this field will be populated with the intermediate secret name, but does not say anything about naming convention.

baijum replied:
Done. 832c72fcec87d9f227ff5ea071ff759a112edc95

pedjak commented on test/acceptance/features/bindAppToServiceUsingSecret.feature:
```diff
@@ -761,3 +761,202 @@ Feature: Bind values from a secret referred in backing service resource
+            """
+
+        Then Service Binding "bind-provisioned-service-4" is ready
+        And Content of file "/bindings/bind-provisioned-service-4/username" in application pod is
```
we should have here the assert about the secret name in the status

baijum replied:
This was done. See line no. 876

pedjak commented on test/acceptance/features/bindAppToServiceUsingSecret.feature:
```diff
@@ -761,3 +761,202 @@ Feature: Bind values from a secret referred in backing service resource
+            """
+
+        Then Service Binding "bind-provisioned-service-1" is ready
+        And jq ".status.secret" of Service Binding "bind-provisioned-service-1" should starts with "bind-provisioned-service-1"
```
we should not advertise this kind of info to users - it is not a part of api/contract - the spec only says that this field will be populated with the intermediate secret name, but does not say anything about naming convention.

baijum replied:
See the last sentence in this section: https://github.com/k8s-service-bindings/spec#service-binding
It reads like this:
>  A Service Binding resource SHOULD reflect the secret projected into the application as `.status.binding.name`.

That's a recommendation.

pedjak replied:
it says only that the secret name should be set in `.status.binding.name`, but it does not force any particular naming convention.

baijum replied:
Done. 832c72fcec87d9f227ff5ea071ff759a112edc95

baijum commented on test/acceptance/features/bindAppToServiceUsingSecret.feature:
```diff
@@ -761,3 +761,202 @@ Feature: Bind values from a secret referred in backing service resource
+                resource: deployments
+              mappings:
+                - name: username_with_password
+                  value: '{{ .username }}:{{ .password }}'
```
@pedjak Please let me know referring to Secret resource values like this is correct.
This seems to be working, but not sure the semantics is right.

When I tried to use the prefix with `data` (e.g., `.sec.data.username`) it gives the base64 encoded value. Is that also expected?

pedjak replied:
Yes, both things work as expected.

baijum commented:
/test 4.7-acceptance

pedjak commented:
@baijum https://github.com/redhat-developer/service-binding-operator/pull/984#pullrequestreview-691947110 is still not addressed.

baijum commented:
> also, we support already using secret as reference, via annotations. Check this scenario:
> 
> https://github.com/redhat-developer/service-binding-operator/blob/master/test/acceptance/features/bindAppToService.feature#L796
> 
> How we are going to differentiate? Should we treat a secret as direct reference only if it does not contain annotations?

@pedjak I have implemented it here:
7903e02544667fd362e723947b55f9c5859131c8

baijum commented:
> > also, we support already using secret as reference, via annotations. Check this scenario:
> > https://github.com/redhat-developer/service-binding-operator/blob/master/test/acceptance/features/bindAppToService.feature#L796
> > How we are going to differentiate? Should we treat a secret as direct reference only if it does not contain annotations?
> 
> @pedjak I have implemented it here:
> [7903e02](https://github.com/redhat-developer/service-binding-operator/commit/7903e02544667fd362e723947b55f9c5859131c8)

I think no need to retry. Then annotations will not work. Changed the approach here:
07f2b439dd1f79bd91721855e7db192b8f0bd093

pedjak commented on pkg/reconcile/pipeline/handler/collect/impl.go:
```diff
@@ -131,6 +132,31 @@ func ProvisionedService(ctx pipeline.Context) {
+		if res.GetKind() == "Secret" && res.GetAPIVersion() == "v1" && res.GroupVersionKind().Group == "" {
+			annotations := res.GetAnnotations()
+			for k := range annotations {
+				if strings.HasPrefix(k, "service.binding") {
```
please `AnnotationPrefix` constant: https://github.com/redhat-developer/service-binding-operator/blob/d680987ea0f2869557ada3f3dcc621981cb1d427/pkg/binding/annotationmapper.go#L34

baijum replied:
Done. c7e29669fb69cbc23bbe07b4309aed4e98c11ba4

pedjak commented on pkg/reconcile/pipeline/handler/collect/impl_test.go:
```diff
@@ -745,6 +746,105 @@ var _ = Describe("Collect From Provisioned Service", func() {
+			collect.DirectSecretReference(ctx)
+		})
+
+		It("continue processing if secret has annotations", func() {
```
stages are arranged in the pipeline, so not sure what do you mean by "continue processing"?  The case is here that a secret will not be considered as direct secret references if it has a service.binding annotation attached to it. Could you please rephrase the test name to reflect that?

baijum replied:
How about something like this?
- `skip processing if the secret has annotations`
- `ensure secret with annotations are ignored`
Please suggest.

pedjak replied:
`Ignore secret having a secret binding annotation`

baijum replied:
Done. 387c9cc3e38705018063de2dea18001a762885d7

baijum commented on test/acceptance/features/bindAppToServiceUsingSecret.feature:
```diff
@@ -761,3 +761,202 @@ Feature: Bind values from a secret referred in backing service resource
+                resource: deployments
+          """
+        Then Service Binding "bind-provisioned-service-1" is ready
+        And jq ".status.secret" of Service Binding "bind-provisioned-service-1" should be changed to "provisioned-secret-1"
```
This needs to verified.

pedjak reviewed (approved):
please squash, otherwise LGTM

baijum commented:
/test 4.7-acceptance

baijum commented:
/test 4.7-acceptance

pedjak commented:
/lgtm

pmacik commented:
/lgtm
/approve
